prevent-broker-scale-down-if-contains-partition-replicas.md

```diff
@@ -0,0 +1,82 @@
+- The `canScaleDownBrokers()` utility method will be present at the top of the compose chain in the `reconcile()` method of the `KafkaReconciler` to make sure that every other method which requires the replica count use the correct replica count based on the outcome of the check.
+- The `canScaleDownBrokers()` method will only run if we see the current Kafka replicas (replicas before the Kafka custom resource is modified) count gets greater than the Kafka replicas present in the Kafka custom resource.
+  We can get the desired Kafka replica count by using `kafka.getReplicas()` where `kafka` is an object of `KafkaCluster` class .
+- This method will check if the broker contains any partition replicas or not and will continue the process based on the outcome.
+- For doing so, the topic metadata will be queried to detect if the broker contains any partition replicas.
+- An Admin client instance will be used to connect with the cluster and get us the topic details(topic name and topic description)
+- Then we can use this information to check if the broker contains any partition replicas or not.
+- The scale down is done after we make sure that the brokers that are going to be removed doesn't contain any partition replicas.
+  By doing this we avoid any partial scale down.
+
+### What to do if a broker contains partitions?
+
+#### Flow:
+
+- If partition replicas are found out on the broker we will revert back the Kafka replicas to the previous count by setting replicas directly in the `KafkaCluster` class using `setReplicas()` method. 
+  Changing the Kafka replica count directly in the Kafka Cluster would help us ensure that we keep same replicas everywhere like while generating certificates, services, ingresses, routes etc.
+- The broker certificates, services, ingresses, routes etc. will be treated with the original number of nodes and the rest of the reconciliation will be done normally.
+- We also generate a new condition which will be added to Kafka resource status depicting that the scale down is not done. It will also contain the `spec.replicas` count(which is being currently being used) in the condition message.
+  ```yaml
+  status:
+    clusterId: DoRj5f84Sruq_7TJ31y7Zw
+    conditions:
+      - lastTransitionTime: "2023-02-22T10:18:56.578009768Z"
+        message: 'Cannot Scale down since broker contains partition replicas. The `spec.kafka.replicas` should be reverted back to 4 directly in the Kafka resource'.
+        reason: ScaleDownException
+        status: "True"
+        type: Warning
+      - lastTransitionTime: "2023-02-22T10:18:57.664668863Z"
+        status: "True"
+        type: Ready
+  ```
+  Note: By the time the replicas are reverted back, the storage validation will be already complete based on the replica count present in Kafka custom resource.
+  This can can cause some issues if someone tries to make some forbidden changes (changes that might not be supported) to the storage during this time frame.
```

Comment:
```suggestion
  This can cause some issues if someone tries to make some forbidden changes (changes that might not be supported) to the storage during this time frame.
```